### PR TITLE
[8.x] Adds class handling for Blade echo statements

### PR DESCRIPTION
**Pull request by lukeraymonddowning**
## What is it?

When working in Blade, I'll often want to always output a class in the same manner. I'll use Brick/Money as the example here, but this can apply to pretty much any class you can think of. 

Brick/Money is a final class, so I can't add a `Stringable` interface to it. Whenever I output `{{ $moneyObject }}`, I would want it to format as `£100.00`, for example. The same could be said for a `Carbon` instance. 

This PR adds a new `Blade::handle()` method that can be placed in the boot method of a Service Provider and allows the user to add intercepting closures for any class. The returned value will be outputted in Blade.

Think of this as the Exception Handler, but for Blade.

## Usage

```php
// AppServiceProvider
Blade::handle(Money::class, fn($object) => $object->formatTo('en_GB'));
Blade::handle(Carbon::class, fn($object) => $object->format('d/m/Y')));
```

```blade
<dl>
    <dt>Total</dt>
    <dd>{{ $product->total }}</dd> <!-- This is a money object, but will be outputted as an en_GB formatted string -->
    <dt>Created at</dt>
    <dd>{{ $product->created_at }}</dd> <!-- This is a Carbon object, but will be outputted as English date format -->
</dl>
```

## How does it work?

This works by wrapping a small piece of logic around php output at the parser level, which means that the object is evaluated just in time and can thus be treated as a standard object throughout the entire application; it is only transformed by the PHP script at the time of output.

Feel free to throw any questions back at me :-)

Thanks for all the hard work guys!

### CODE REVIEW

**Comment by lukeraymonddowning** [comment]
This is a fixed version of #37477 which had broken unit tests. It also is faster than previously because it will completely skip the JIT check if no handlers have been specified.

**Comment by taylorotwell** [comment]
Does putting line breaks in the compiled echo syntax cause any unwanted side effects in regards to HTML spacing?

**Comment by lukeraymonddowning** [comment]
Will check this tomorrow and add unit tests to prove

**Comment by lukeraymonddowning** [comment]
@taylorotwell from testing in a real application, the whitespaces made no difference, but it doesn't really make much sense to keep the whitespaces there anyway so I've removed them. Also added a test in similar vain to those in the `BladeEchoTest` file that check for the preservation of newlines after the output.

File: src/Illuminate/View/Compilers/BladeCompiler.php [hunk 1]
```diff
@@ -752,4 +759,14 @@ public function withoutComponentTags()
     {
         $this->compilesComponentTags = false;
     }
+
+    /**
+     * Add a handler to be executed before echoing a given class.
+     *
+     * @return void
+     */
+    public function handle(string $className, callable $handler)
```
Comment (driesvints) [inline src/Illuminate/View/Compilers/BladeCompiler.php:768]: Wondering if this should be `addEchoHandler`. `handle` is so generic and might be confused with a job's `handle` method. The former is a registration of a resolver while the later is an actual resolver.

Comment (lukeraymonddowning) [reply src/Illuminate/View/Compilers/BladeCompiler.php]: Agreed. I've renamed.

**Comment by taylorotwell** [comment]
Renamed method to `stringable` and added `ReflectsClosures` trait to Blade compiler so that it is now possible to omit the class name as the first argument and just provide a type-hinted callback:

```php
Blade::stringable(fn (Money $money) => $money->formatTo('en_GB'));
```

**Comment by lukeraymonddowning** [comment]
Love it @taylorotwell. Stringable fits in perfectly with the interface, so makes complete sense and the ability to type hint instead is pure fire 😎

**Pull request merged**